File: phase-2-persistence/2.3-round-trip-tests/starter/Kingdom.Engine/Buildings/Mine.cs
namespace Kingdom.Engine.Buildings;

public class Mine : Building
{
    public Mine(string name) : base(name) { }
    public Mine(string name, int level) : base(name, level) { }

    public override void Tick(ResourceLedger ledger)
    {
        ledger.Add(Resource.Stone, 2 * Level);
    }
}

File: phase-5-roblox/5.3-oop-in-luau/starter/roblox-kingdom/Engine/Mine.lua
local Building = require(script.Parent.Building)

local Mine = setmetatable({}, { __index = Building })
Mine.__index = Mine

function Mine.new(name: string)
    local self = Building.new(name)
    setmetatable(self, Mine)
    return self
end

function Mine:tick(ledger: any)
    ledger:add("Stone", 2 * self.level)
end

return Mine
